File: inform/Project/Toolbar/IFToolbarProgressIndicator.m
//
//  IFToolbarProgressIndicator.m
//  Inform
//
//  Created by Toby Nelson, 2014.
//

#import "IFToolbarProgressIndicator.h"
#import "IFUtility.h"

@implementation IFToolbarProgressIndicator

- (instancetype)init {
    self = [super init];
    if( self ) {
        // Use core animation, so we get to draw everything (including the current progress) in drawRect.
        [self setWantsLayer:YES];
    }
    return self;
}

- (void)drawRect:(NSRect)dirtyRect {
    NSRect rect = [self bounds];
    CGFloat radius = floor(rect.size.height / 2);
    NSBezierPath *bz = [NSBezierPath bezierPathWithRoundedRect: rect xRadius: radius yRadius: radius];
    
    // Draw progress inside
    [bz setClip];
    rect.size.width = floor(rect.size.width * ([self doubleValue] / [self maxValue]));
    [[NSColor colorNamed:@"StatusIndicator"] set];
    NSRectFill(rect);

    // Draw border
    [bz setLineWidth:1.0];
    [[NSColor colorNamed:@"StatusIndicatorBorder"] set];
    [bz stroke];
}

@end

File: inform/Project/Toolbar/IFToolbarProgressIndicator.swift
//
//  IFToolbarProgressIndicator.swift
//  Inform
//
//  Created by Toby Nelson on 28/09/2023.
//

import Foundation

class IFToolbarProgressIndicator: NSProgressIndicator {

    required init?(coder: NSCoder) {
        super.init(coder: coder)
        self.wantsLayer = true
    }

    override func draw(_ dirtyRect:NSRect) {
        var rect:NSRect = self.bounds
        let radius:CGFloat = floor(rect.size.height / 2)
        let bz:NSBezierPath! = NSBezierPath(roundedRect: rect, xRadius: radius, yRadius: radius)

        // Draw progress inside
        bz.setClip()
        rect.size.width = floor(rect.size.width * (self.doubleValue / self.maxValue))
        NSColor(named: "StatusIndicator")?.set()
        rect.fill()

        // Draw border
        bz.lineWidth = 1.0
        NSColor(named: "StatusIndicatorBorder")?.set()
        bz.stroke()
    }
}
